crates/matrix-sdk-ui/src/timeline/mod.rs

```diff
@@ -331,6 +347,146 @@ impl Timeline {
         self.inner.update_event_send_state(&txn_id, send_state).await;
     }
 
+    /// Toggle a reaction on an event
+    ///
+    /// Adds or redacts a reaction based on the state of the reaction at the
+    /// time it is called.
+    ///
+    /// When redacting an event, the redaction reason is not sent.
+    ///
+    /// Ensures that only one reaction is sent at a time to avoid race
+    /// conditions and spamming the homeserver with requests.
+    ///
+    /// FIXME: If called multiple times in quick succession on the same
+    /// reaction, the function will drop requests until the previous request
+    /// is complete.
+    pub async fn toggle_reaction(&self, annotation: &Annotation) -> Result<(), Error> {
+        let txn_id = TransactionId::new();
+        let user_id = self.user_id()?;
+
+        // Check if the reaction already exists
+        let find_related_event = || async {
+            self.item_by_event_id(&annotation.event_id).await.ok_or(Error::FailedTogglingReaction)
+        };
+
+        let related_event = find_related_event().await?;
+
+        let user_reactions = related_event
+            .reactions()
+            .by_key(&annotation.key)
+            .map(|group| group.by_sender(&user_id));
+
+        // Find an existing reaction to redact, if any
+        let to_redact = user_reactions
+            .map(|mut reactions| reactions.find_map(|(_txid, event_id)| event_id))
+            .unwrap_or(None);
+
+        // Wait for any ongoing reaction requests to finish
+        let lock = self.reaction_lock.lock().await;
+
+        // Check the event again, in case it has changed since we last checked
+        let related_event = find_related_event().await?;
+        let user_reactions = related_event
+            .reactions()
+            .by_key(&annotation.key)
+            .map(|group| group.by_sender(&user_id));
+
+        match to_redact {
+            Some(event_id) => {
+                if user_reactions.is_none() {
+                    // Reaction already redacted or doesn't exist
+                    return Ok(());
+                }
+
+                self.redact_reaction(annotation, event_id, txn_id.clone()).await?
+            }
+            None => {
+                if user_reactions.is_some() {
+                    // Reaction already exists
+                    return Ok(());
+                }
+                self.add_reaction(annotation, txn_id.clone()).await?
+            }
+        };
+
+        // Release only after the request is complete
+        drop(lock);
```

Comment:
This is unnecessary, returning from the function drops the lock automatically.

Reply:
[`4e0393c`](https://github.com/matrix-org/matrix-rust-sdk/pull/2118/commits/4e0393c877b1cb16444bfa52223c3e685205a10c)